File: BOJ/백트래킹/9095.py
# 9095 1, 2, 3 더하기
# https://www.acmicpc.net/problem/9095
import sys
def input(): return sys.stdin.readline().rstrip()


T = int(input())


def recur(n):
    if n == 1:
        return 1
    if n == 2:
        return 2
    if n == 3:
        return 4
    else:
        return recur(n-1) + recur(n-2) + recur(n-3)


for _ in range(T):
    n = int(input())
    result = recur(n)
    print(result)

File: BOJ/DP/9095.cpp
// 9095 1,2,3 더하기
// https://www.acmicpc.net/problem/9095
#include <iostream>

using namespace std;

int main(void) {
    int T, N;

    // 1 : 1
    // 2 : 1+1, 2 -> 2
    // 3 : 1+1+1, 2+1, 1+2, 3 -> 4

    // 4: 1+1+1+1, 1+1+2, 1+2+1, 2+1+1, 2+2, 1+3, 3+1 -> 7
    // 5: 13
    // 6 24
    // 7 44
    int *dp = new int[12];
    dp[1] = 1;
    dp[2] = 2;
    dp[3] = 4;

    for (int i = 4; i < 12; ++i)
    {
        dp[i] = dp[i-1]+dp[i-2]+dp[i-3];
    }

    cin >> T;
    for (int t = 0; t < T; ++t)
    {
        cin >> N;
        cout << dp[N] << '\n';
    }

    delete[] dp;
    return 0;
}